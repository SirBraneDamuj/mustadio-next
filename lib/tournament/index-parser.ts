import { JSDOM } from "jsdom";

type DumpFile = {
  name: string;
  timestamp: string;
};

type IndexHtml = {
  latestTournament: string;
  dumpFiles: DumpFile[];
};

export function loadIndexHtml(html: string): IndexHtml {
  const { window } = new JSDOM(html);
  const document = window.document;
  let latestTournament = "";
  const dumpFiles = [];
  const linkNodes = document.querySelectorAll("tr > td > a");
  for (const node of linkNodes) {
    const href = node.getAttribute("href");
    if (!href) {
      continue;
    }
    if (href.startsWith("tournament_")) {
      if (href > latestTournament) {
        latestTournament = href;
      }
    } else if (href.endsWith(".txt")) {
      const timestamp =
        node.parentNode?.nextSibling?.textContent?.trim() || "no timestamp";
      dumpFiles.push({
        name: href,
        timestamp,
      });
    }
  }
  return {
    dumpFiles,
    latestTournament: latestTournament.slice(0, -1),
  };
}
