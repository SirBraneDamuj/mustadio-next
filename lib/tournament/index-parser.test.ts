import { promises as fs } from "fs";
import path from "path";
import { describe, expect, it } from "vitest";
import { loadIndexHtml } from "./index-parser";

describe("loadIndexHtml", () => {
  it("parses index.html correctly (snapshot)", async () => {
    const filePath = path.join(
      __dirname,
      "../../resources/fftbg_fake/index.html"
    );
    const html = await fs.readFile(filePath, "utf-8");
    const result = loadIndexHtml(html);

    expect(result).toMatchSnapshot();
  });
});
