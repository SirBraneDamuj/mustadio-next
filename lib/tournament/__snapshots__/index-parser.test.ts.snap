// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadIndexHtml > parses index.html correctly (snapshot) 1`] = `
{
  "dumpFiles": [
    {
      "name": "BlueSkills.txt",
      "timestamp": "2020-02-06 19:53",
    },
    {
      "name": "BlueSkillsets.txt",
      "timestamp": "2020-02-06 19:50",
    },
    {
      "name": "MonsterSkills.txt",
      "timestamp": "2019-12-10 14:01",
    },
    {
      "name": "Monsters.txt",
      "timestamp": "2019-12-11 09:28",
    },
    {
      "name": "PrestigeSkills.txt",
      "timestamp": "2020-02-05 08:28",
    },
    {
      "name": "UserSkills.txt",
      "timestamp": "2020-01-26 12:46",
    },
    {
      "name": "classhelp.txt",
      "timestamp": "2020-02-06 16:39",
    },
    {
      "name": "highexp.txt",
      "timestamp": "2020-02-06 23:59",
    },
    {
      "name": "highscores.txt",
      "timestamp": "2020-02-06 23:59",
    },
    {
      "name": "infoability.txt",
      "timestamp": "2020-02-06 11:49",
    },
    {
      "name": "infoitem.txt",
      "timestamp": "2020-02-06 13:04",
    },
    {
      "name": "infostatus.txt",
      "timestamp": "2020-02-01 19:02",
    },
    {
      "name": "skillhelp.txt",
      "timestamp": "2020-01-29 13:38",
    },
    {
      "name": "stats.txt",
      "timestamp": "2020-02-06 23:59",
    },
    {
      "name": "zodiachelp.txt",
      "timestamp": "2020-01-08 08:19",
    },
  ],
  "latestTournament": "tournament_1581051562581",
}
`;
